compiler/src/openxla/compiler/nvgpu/Transforms/ConvertMHLOToCUDNN.cpp

```diff
@@ -0,0 +1,94 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <algorithm>
+#include <numeric>
+
+#include "mhlo/IR/hlo_ops.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "openxla/compiler/nvgpu/Dialect/CUDNN/IR/CUDNNDialect.h"
+#include "openxla/compiler/nvgpu/Dialect/CUDNN/IR/CUDNNOps.h"
+#include "openxla/compiler/nvgpu/Transforms/Passes.h"
+
+#define GEN_PASS_DEF_CONVERTMHLOTOCUDNN
+#include "openxla/compiler/nvgpu/Transforms/Passes.h.inc"
+
+using namespace mlir;
+
+namespace openxla::compiler::nvgpu {
+
+static SmallVector<int64_t> getRowMajorStrides(ArrayRef<int64_t> shape) {
+  SmallVector<int64_t> strides(shape.size(), 1);
+  std::partial_sum(shape.rbegin(),
+                   shape.rend() - 1,
+                   strides.rbegin() + 1,
+                   std::multiplies<int64_t>());
+  return strides;
+}
+
+static cudnn::TensorDescType getTensorDescType(TensorType tensor_type) {
+  auto shape = tensor_type.getShape();
+  Type element_type = tensor_type.getElementType();
+  int alignment = 0;
+  auto strides = getRowMajorStrides(shape);
+  return cudnn::TensorDescType::get(tensor_type.getContext(),
+                                    shape, element_type, alignment, strides);
+}
+
+namespace {
+
+struct ConvertClamp : public OpRewritePattern<mhlo::ClampOp> {
+  using OpRewritePattern<mhlo::ClampOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(mhlo::ClampOp op,
+                                PatternRewriter& rewriter) const override {
+    llvm::APFloat min;
+    if (!matchPattern(op.getMin().getDefiningOp(), m_ConstantFloat(&min))) {
+      return rewriter.notifyMatchFailure(op, "expected constant min");
+    }
+    llvm::APFloat max;
+    if (!matchPattern(op.getMax().getDefiningOp(), m_ConstantFloat(&max)) ||
+        !max->isNan()) {
+      return rewriter.notifyMatchFailure(op, "expected NaN max");
+    }
+    TensorType tensor_type = op.getOperand().getType();
+    cudnn::TensorDescType tensor_desc_type = getTensorDescType(tensor_type);
+    Value input = rewriter.create<UnrealizedConversionCastOp>(
```

Comment:
So here you are using the descriptor and the value as being castable, but the descriptor is the metadata of the lowered tensor (well, it is more the metadata of the memref really as these are in buffered world with explicit layouts and backing memory represented). Doesn't mean we can't change that, just considered this more of post bufferization level. But good to check here what folks are thinking.

Reply:
I didn't intend to imply these being castable. I wouldn't read anything into what the pass does, it's really just a placeholder.